include/deal.II/base/partitioner.templates.h

```diff
@@ -0,0 +1,439 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2017 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_partitioner_templates_h
+#define dealii_partitioner_templates_h
+
+#include <deal.II/base/config.h>
+#include <deal.II/base/partitioner.h>
+#include <deal.II/lac/la_parallel_vector.h>
+
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace Utilities
+{
+  namespace MPI
+  {
+
+#ifndef DOXYGEN
+
+#ifdef DEAL_II_WITH_MPI
+
+    template <typename Number>
+    void
+    Partitioner::export_to_ghosted_array_start(const unsigned int             communication_channel,
+                                               const ArrayView<const Number> &locally_owned_array,
+                                               const ArrayView<Number>       &temporary_storage,
+                                               const ArrayView<Number>       &ghost_array,
+                                               std::vector<MPI_Request>      &requests) const
+    {
+      AssertDimension(locally_owned_array.size(), local_size());
+      AssertDimension(temporary_storage.size(), n_import_indices());
+      Assert(ghost_array.size() == n_ghost_indices() ||
+             ghost_array.size() == n_ghost_indices_in_larger_set,
+             ExcMessage(std::string("The size of the ghost index array (")
```

Comment:
you use it below in 3 other places, i would simply declare a single exception via `DeclException3` and use it.